docs/weave-tutorial/07-handlers.md

```diff
@@ -0,0 +1,675 @@
+}
+```
+
+`CreateCommentHandler` saves the a comment that is posted to an article if all conditions are met. `validation` method step by step:
+
+1. Message is validated with static checks such as missing title check.
+2. `ArticleBucket` is queried if the article that is desired to be commented exists.
+3. `Comment` instance is created with tx signers address and creation time. If you want to check if signer is eligible to executed a transaction, you can compare it with the owner data that is kept in bucket and if there is no match then tx will be aborted and the fee will be deduced from transaction sender as anti spam fee.
+
+```go
+// Deliver creates a comment and saves if all preconditions are met
+func (h CreateCommentHandler) Deliver(ctx weave.Context, store weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+    _, comment, err := h.validate(ctx, store, tx)
+    if err != nil {
+        return nil, err
+    }
+
+    err = h.cb.Put(store, comment)
+    if err != nil {
+        return nil, errors.Wrap(err, "cannot store comment")
+    }
+
+    // Returns generated user ID as response
+    return &weave.DeliverResult{Data: comment.ID}, nil
+}
+```
+
+As you noticed most of the hard work done in `validate` method and `Deliver` methods are used to handle database related work.
+
+Once `validate` is implemented, `Check` must ensure it is valid
+and then return a rough cost of the message, which may be based
+on the storage cost of the text of the post. This return value
+is similar to the concept of _gas_ in ethereum, although it doesn't
+count to the fees yet, but rather is used by tendermint to prioritize
+the transactions to fit in a block.
+
+Here is an example of more dynamic fee deduction:
+
+```go
+// Check just verifies it is properly formed and returns
+// the cost of executing it.
+func (h CreateArticleHandler) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+    msg, _, err := h.validate(ctx, store, tx)
+    if err != nil {
+        return nil, err
+    }
+
+    // Calculate gas cost
+    gasCost := int64(len(msg.Content)) * newArticleCost / articleCostUnit
+
+    return &weave.CheckResult{GasAllocated: gasCost}, nil
+}
+```
+
+In the case of an Article creation, we decided to charge the author 1 gas per mile characters with the first 1000 characters offered: `gasCost := int64(len(msg.Content)) * newArticleCost / articleCostUnit`
+
+## Testing Handlers
+
+In order to test a handler, we need four things :
+
+- A storage
+- A weave context
+- An Authenticator associated with our context
+- A Tx object to process (eg. to check or to deliver)
+
+There is a ready to use in memory storage available in the [store package](https://github.com/iov-one/weave/blob/master/store/btree.go#L31-L36). There are also util functions available that we can use to create a weave context with a list of signers (eg. authorized addresses) via an [Authenticator](weave/design/permissions)
```

Comment:
`There is a ready-to-use in memory storage available in the [store package](https://github.com/iov-one/weave/blob/master/store/btree.go#L31-L36). There are also util functions available that we can use to create a weave context with a list of signers (eg. authorized addresses) via an [Authenticator](weave/design/permissions)
`